Optimize integer-->string conversions

kimikage opened the pull request:
~This avoids invalidations caused by invalidating `StringVector(::Integer)`. (cf. PR #36459)~ (**Edit:** The problem has been solved by PR #36459.)
This also makes `bin()`, `dec()` and `hex()` slightly faster, but does not change the Printf.

```julia
julia> versioninfo() # on WSL
Julia Version 1.6.0-DEV.312
Commit 6185d24ac3 (2020-06-28 22:42 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)

julia> @btime string.(x, base=2) setup=(x=rand(Int64,1000));
  57.300 μs (2001 allocations: 151.67 KiB) # before
  38.400 μs (2001 allocations: 151.69 KiB) # after

julia> @btime string.(x, base=8) setup=(x=rand(Int64,1000));
  34.700 μs (2001 allocations: 117.30 KiB) # before
  34.800 μs (2001 allocations: 117.30 KiB) # after

julia> @btime string.(x, base=10) setup=(x=rand(Int64,1000));
  45.900 μs (2001 allocations: 117.27 KiB) # before
  35.400 μs (2001 allocations: 117.28 KiB) # after

julia> @btime string.(x, base=10) setup=(x=rand(Int32,1000));
  35.300 μs (2001 allocations: 101.69 KiB) # before
  31.300 μs (2001 allocations: 101.69 KiB) # after

julia> @btime string.(x, base=16) setup=(x=rand(Int64,1000));
  38.300 μs (2001 allocations: 115.84 KiB) # before
  33.600 μs (2001 allocations: 115.80 KiB) # after
```

kimikage committed:
Optimize integer-->string conversions

This avoids invalidations caused by invalidating `StringVector(::Integer)`.
This also makes `bin()`, `dec`() and `hex()` slightly faster,
but does not change the Printf.

JeffBezanson commented:
Awesome!

StefanKarpinski commented on base/intfuncs.jl:
```diff
@@ -602,74 +602,111 @@ ndigits(x::Integer; base::Integer=10, pad::Integer=1) = max(pad, ndigits0z(x, ba
-    i = neg + max(pad,sizeof(x)<<3-leading_zeros(x))
-    a = StringVector(i)
+    m = 8sizeof(x) - leading_zeros(x)::Int
+    n = neg + max((pad % Int)::Int, m)
```
Is the type assertion really necessary in `(pad % Int)::Int`? That seems like something the compiler should know.

timholy replied:
> Is the type assertion really necessary in (pad % Int)::Int

In some circumstances, yes. All applicable methods infer to `Int` but there are 14 of them:

```julia
julia> sig = (Integer, Type{Int})
(Integer, Type{Int64})

julia> cts = code_typed(%, sig);

julia> collect(zip(methods(%, sig), last.(cts)))
14-element Array{Tuple{Method,DataType},1}:
 (rem(x::BigInt, ::Type{T}) where T<:Union{Int16, Int32, Int64, Int8, UInt16, UInt32, UInt64, UInt8} in Base.GMP at gmp.jl:332, Int64)
 (rem(x::BigInt, ::Type{T}) where T<:Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8} in Base.GMP at gmp.jl:335, Int64)
 (rem(x::UInt32, ::Type{Int64}) in Base at int.jl:479, Int64)
 (rem(x::Int8, ::Type{Int64}) in Base at int.jl:477, Int64)
 (rem(x::Bool, ::Type{Int64}) in Base at int.jl:474, Int64)
 (rem(x::UInt16, ::Type{Int64}) in Base at int.jl:479, Int64)
 (rem(x::UInt128, ::Type{Int64}) in Base at int.jl:472, Int64)
 (rem(x::UInt64, ::Type{Int64}) in Base at int.jl:482, Int64)
 (rem(x::UInt8, ::Type{Int64}) in Base at int.jl:479, Int64)
 (rem(x::Int128, ::Type{Int64}) in Base at int.jl:472, Int64)
 (rem(x::Int32, ::Type{Int64}) in Base at int.jl:477, Int64)
 (rem(x::Int16, ::Type{Int64}) in Base at int.jl:477, Int64)
 (rem(x::T, ::Type{T}) where T<:Integer in Base at int.jl:540, Int64)
 (rem(x::Integer, T::Type{var"#s71"} where var"#s71"<:Integer) in Base at int.jl:543, Int64)
```
and `max_methods` is 3. So the abstractly-inferred call is judged to have a return type of `Any`, even if any concrete type is inferred correctly:
```julia
julia> f(x) = x % Int
f (generic function with 1 method)

julia> code_typed(f, (Int,))
1-element Array{Any,1}:
 CodeInfo(
1 ─     return x
) => Int64

julia> code_typed(f, (Integer,))
1-element Array{Any,1}:
 CodeInfo(
1 ─ %1 = (x % Main.Int)::Any
└──      return %1
) => Any
```

Because we are discovering more and more good reasons *not* to run inference to exhaustion (and developing more infrastructure to limit inference, like `@nospecialize`), this is becoming more of an issue. This is basically what #36463 is about.

StefanKarpinski replied:
Gotcha. Carry on.

timholy reviewed (approved):
This is a pretty performance-sensitive operation, so unrolling those loops is a great idea.

rfourquet commented on base/intfuncs.jl:
```diff
@@ -602,74 +602,111 @@ ndigits(x::Integer; base::Integer=10, pad::Integer=1) = max(pad, ndigits0z(x, ba
-    i = neg + max(pad,sizeof(x)<<3-leading_zeros(x))
-    a = StringVector(i)
+    m = 8sizeof(x) - leading_zeros(x)::Int
+    n = neg + max((pad % Int)::Int, m)
```
Using `pad % Int` is invalid in some corner cases (e.g. for `pad == typemax(Int)+1`). It's not necessarily a problem for `bin` as it's an internal function, but some form of validation should probably remain somewhere, e.g. in client functions like `string` (which could take care of converting to `Int`).

kimikage replied:
Is that a security issue?

rfourquet replied:
I don't know, but this makes the behavior of `string` sometimes incorrect, in an  and observable way.

kimikage replied:
> an observable way

Can you give a specific example? Being secure is a requirement of the correctness, but otherwise I don't think the correct behavior is obvious.

rfourquet replied:
Currently, `string(0, pad=typemin(Int128)+3)` returns `""`, which seems correct as per the docstring. With this change (if I'm not mistaken), it will return `"000"`, which seems wrong.

kimikage replied:
I will try to fix it. However, this PR starts with avoiding invalidation in the first place.

rfourquet replied:
> However, this PR starts with avoiding invalidation in the first place.

I don't understand how this is related. I'm not familiar with invalidations hunting, but hopefully we can both avoid invalidations and be correct. In this case, what about `pad = clamp(pad, Int)::Int` ?

kimikage replied:
It's "usually" not a bad idea. However, I don't really want to limit it to the range of `Int` (since it is too "wide" for practical use cases). I want to just avoid the invalidation. 
The `pad`s are always `Int`, in the use cases in `Base` i.e. in the default sysimage. Therefore, `pad % Int` is just insurance. In that respect, I think `clamp` is not better than `rem`.

Perhaps a reasonable fix would be to get rid of the `% Int` (i.e. reverting). Do you have any thoughts?

rfourquet replied:
> I don't really want to limit it to the range of Int (since it is too "wide" for practical use cases)

I don't really understand this sentence. In my view, everything which is `<0` can be set to `0`, and everything which is bigger than `typemax(Int)` probably makes no sense has it's not practical to have a string of length `> typemax(Int)`, so such number can be replaced by `typemax(Int)`. I just described `clamp`. What doesn't seem OK is to use `pad % Int` as this has a weird behavior of for these extreme values of `pad`.

> The pads are always Int, in the use cases in Base i.e. in the default sysimage. 

Maybe, but I'm concerned by usage outside of base where your current change leads to an incorrect result as I showed with an example. But unfortunately I have no other suggestion, as I didn't follow how invalidation work, and I don't know which is better, `clamp` or the `max` that was there or writing the equivalent of `clamp` by hand (which would be sad) etc.

kimikage replied:
>> I don't really want to limit it to the range of Int (since it is too "wide" for practical use cases)
>
> I don't really understand this sentence.

I meant a user friendly error message or docstring.

kimikage committed:
Change `pad` option handling

kimikage commented:
@rfourquet, I appreciate your comment

I added the clamping into the `string()`, and changed the types of the `pad` of internal functions (e.g. `bin()`) from `Integer` to `Int`. Although we can leave them as `Integer` for "compatibility", I think there are no practical advantages.

I don't know the reason, but there is a little slowdown (~1 ns/number) from the original commit, but the unrolling is still effective.

timholy commented:
Bumping this. I don't have any objections to merging this.

kimikage commented:
Since PR #36459 was merged, the `::Int` for `leading_zeros(x)` doesn't seem to be needed anymore.

kimikage committed:
Omit unnecessary type assertions

kimikage committed:
Add out-of-range `base` tests

musm commented:
Looks like this PR has been approved, is it ok to merge?

musm commented on base/intfuncs.jl:
```diff
@@ -601,75 +601,112 @@ ndigits(x::Integer; base::Integer=10, pad::Integer=1) = max(pad, ndigits0z(x, ba
-    i = neg + max(pad,sizeof(x)<<3-leading_zeros(x))
-    a = StringVector(i)
+function bin(x::Unsigned, pad::Int, neg::Bool)
+    m = 8sizeof(x) - leading_zeros(x)
```
```suggestion
    m = sizeof(x) - leading_zeros(x)
```

musm replied:
for clarity, I think this is an improvement

musm commented on base/intfuncs.jl:
```diff
@@ -601,75 +601,112 @@ ndigits(x::Integer; base::Integer=10, pad::Integer=1) = max(pad, ndigits0z(x, ba
-    i = neg + max(pad,div((sizeof(x)<<3)-leading_zeros(x)+2,3))
-    a = StringVector(i)
+function oct(x::Unsigned, pad::Int, neg::Bool)
+    m = div(8sizeof(x) - leading_zeros(x) + 2, 3)
```
```suggestion
    m = div(sizeof(x) - leading_zeros(x) + 2, 3)
```

musm replied:
Any reason the tests did not catch this?

musm commented on base/intfuncs.jl:
```diff
@@ -601,75 +601,112 @@ ndigits(x::Integer; base::Integer=10, pad::Integer=1) = max(pad, ndigits0z(x, ba
-        x >>= 4
-        i -= 1
+function hex(x::Unsigned, pad::Int, neg::Bool)
+    m = 2sizeof(x) - (leading_zeros(x) >> 2)
```
```suggestion
    m = 2*sizeof(x) - (leading_zeros(x) >> 2)
```

kimikage committed:
Modify multiplication notation

StefanKarpinski commented:
I've kicked off the failed CI runs, which looked unrelated. This seems good once CI passes.

kimikage commented:
Thanks!

kimikage commented:
I'm sorry, but I may have broken something.
cc: @rfourquet
https://github.com/JuliaCI/NanosoldierReports/blob/c1103cd2398e1d0442f115264893823991742a66/pkgeval/by_date/2020-10/01/logs/BitIntegers/1.6.0-DEV-2fe2b43606.log

rfourquet commented:
> Perhaps the cause is the type of bit shifting.

@kimikage Thank you for the report, I had no idea this `x << 0x4` was broken in `BitIntegers`.
